added v1alpha1 API for PodPreset under settings

droot opened the pull request:
This PR adds `v1alpha1` PodPreset API under `settings` group.

Sending this for an early review since this is my first contribution to Service Catalog repo.

Note for Reviewers:
I am working on the tests. This API will be guarded by a flag and will be disabled by default.

MHBauer commented on Makefile:
```diff
@@ -189,10 +189,42 @@ $(BINDIR)/e2e.test: .init
 		--v 1 --logtostderr \
 		--go-header-file "vendor/github.com/kubernetes/repo-infra/verify/boilerplate/boilerplate.go.txt" \
-		--input-dirs "github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1alpha1,k8s.io/client-go/pkg/api/v1,k8s.io/apimachinery/pkg/apis/meta/v1" \
+		--input-dirs "github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1alpha1,github.com/kubernetes-incubator/service-catalog/pkg/apis/settings/v1alpha1,k8s.io/client-go/pkg/api/v1,k8s.io/apimachinery/pkg/apis/meta/v1" \
```
Maybe a TODO For later, Can we try splitting up `--input-dirs` args like above?

I can't remember if I did or did not test this with openapi-gen.

droot replied:
No need for TODO :). Fixed it. Thanks for the suggestion.

MHBauer commented on cmd/apiserver/apiserver.go:
```diff
@@ -26,11 +26,13 @@ import (
+	// API server process.
 	// Please do not remove.
 	_ "github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/install"
+	_ "github.com/kubernetes-incubator/service-catalog/pkg/apis/settings/install"
```
character flaw: I pretty much can't see these and not comment on them. `init()`s, yug.

droot replied:
haha .. I know init()s. Can't help it.

MHBauer commented:
Looks like a good start. I cannot comment on the changes needed for TPRs.

pmorie commented:
I suggest that we do not make TPR support for this API group a precondition for merging this PR

droot commented:
I will split the PR into two commits. One capturing the human edits and another one capturing generated files, so that it is easy to review.

droot commented:
@MHBauer @pmorie PR has two separate commits one for manual and other for generated files. This should help in the review. Also added generated files for client which was not part of the original commit. PTAL whenever you get a chance.

About TPR, I have added the plumbing, but will not be testing TPR support for this API group.

MHBauer commented:
Having PodPresets in this server seems to blow up TPR support completely. If service-catalog plans to rely on PodPresets, and PodPresets rely on etcd, etcd access must always be available.

Not sure what to think about this. Even if it wasn't in our API Server and was instead in a separate second API Server, it would still need to be installed with etcd support, which in the TPR case, it could not get.

droot commented:
@MHBauer Re: TPR/Etcd, I would propose to go ahead with Etcd only support for now because it is going to be feature guarded anyways. And then add CRD support later when we decide to implement CRD support for all the resources for service catalog. Once it is backed by CRD support, we can enable PodPreset by default. What do you think ?

MHBauer commented:
Fine with me. Want to make sure the people who have expectations of CRD support understand that it isn't going to exist to start. 

As far as tests, on the integration level it should be easy enough to no-op the TPR/CRD test as they are separated as subtests.

MHBauer commented on pkg/registry/settings/podpreset/storage.go:
```diff
@@ -0,0 +1,156 @@
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/client-go/pkg/api"
+	"k8s.io/kubernetes/pkg/api"
```
this won't compile. api imported twice. api and api imported again.

MHBauer commented on Makefile:
```diff
@@ -189,11 +189,15 @@ $(BINDIR)/e2e.test: .init
+		--input-dirs "k8s.io/client-go/pkg/api/v1" \
+		--input-dirs "k8s.io/client-go/pkg/api" \
+		--input-dirs "k8s.io/apimachinery/pkg/apis/meta/v1" \
+		--output-package "github.com/kubernetes-incubator/service-catalog/pkg/openapi"
```
does this require to be moved up here? I don't think there are any dependencies in either direction, but I'd rather not move things around if they don't need it.

droot commented on charts/catalog/values.yaml:
```diff
@@ -72,7 +72,7 @@ apiserver:
 controllerManager:
   # controller-manager image to use
-  image: quay.io/kubernetes-service-catalog/controller-manager:v0.0.14
+  image: localhost:6000/controller-manager:canary
```
This file will not be committed.

droot commented on pkg/apis/settings/v1alpha1/register.go:
```diff
@@ -0,0 +1,55 @@
+)
+
+// GroupName is the group name use in this package
+const GroupName = "settings.k8s.io"
```
@pmorie need help here to figure out what group name to use. In current form, it will conflict with settings group  in core kubernetes.

pmorie replied:
`settings.servicecatalog.k8s.io` for now, maybe?

droot commented:
@MHBauer @pmorie added integration tests for the PodPreset APIs. This is very close to final changes (need to remove TPR support). PTAL.

droot commented:
Removed TPR support as suggested by @pmorie .

droot commented:
@MHBauer @pmorie Guarded the PodPreset APIs using feature gates. The PR have addressed all of your previous comments. PTAL.

MHBauer commented on pkg/registry/settings/podpreset/storage.go:
```diff
@@ -0,0 +1,125 @@
+		DestroyFunc: dFunc,
+	}
+
+	// TODO: PodPreset doesn't have any Status field, so statusStore is not
```
let's not put it in if we don't need it.

droot replied:
Done.

MHBauer commented on pkg/registry/settings/rest/storage_settings.go:
```diff
@@ -0,0 +1,123 @@
+
+	storage := map[string]rest.Storage{}
+	if apiResourceConfigSource.ResourceEnabled(version.WithResource("podpresets")) {
+		podPresetStorage, _, err := podpreset.NewStorage(*podPresetOpts)
```
let's get rid of status.

droot replied:
Done.

pmorie commented on Makefile:
```diff
@@ -192,8 +192,37 @@ $(BINDIR)/e2e.test: .init
+		--input-dirs "k8s.io/apimachinery/pkg/apis/meta/v1" \
+		--input-dirs "k8s.io/api/core/v1" \
 		--output-package "github.com/kubernetes-incubator/service-catalog/pkg/openapi"
+	# Generate defaults
```
I think a comment delineating that the remaining generator calls are for the settings API would be helpful

pmorie commented on build/update-client-gen.sh:
```diff
@@ -26,28 +26,32 @@ BINDIR=${REPO_ROOT}/bin
 ${BINDIR}/client-gen "$@" \
 	      --input-base "github.com/kubernetes-incubator/service-catalog/pkg/apis/" \
-	      --input servicecatalog/ \
+	      --input servicecatalog/,settings/ \
```
is it possible to use distinct input operations here?

pmorie commented on charts/catalog/templates/controller-manager-deployment.yaml:
```diff
@@ -77,7 +77,7 @@ spec:
             path: /healthz
           failureThreshold: 1
-          initialDelaySeconds: 10
+          initialDelaySeconds: 20
```
did you mean to commit this?

pmorie replied:
If so might be useful as a separate PR which could merge faster than this one :)

pmorie commented on charts/catalog/templates/apiserver-deployment.yaml:
```diff
@@ -58,6 +58,8 @@ spec:
         - --disable-auth
         {{- end }}
+        - --feature-gates
+        - {{ .Values.apiserver.featureGates }}
```
Seems like it might be good to wrap this in something like `if .Values.setFeatureGates`

droot replied:
Done.

pmorie commented on pkg/apis/settings/types.go:
```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
2017

droot replied:
Done

pmorie commented on pkg/apis/settings/v1alpha1/types.go:
```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Check dates in these files :)

droot replied:
Done :)

pmorie commented on pkg/apis/settings/validation/validation.go:
```diff
@@ -0,0 +1,69 @@
+		allErrs = append(allErrs, field.Required(fldPath.Child("volumes", "env", "envFrom", "volumeMounts"), "must specify at least one"))
+	}
+
+	// volumes, vErrs := apivalidation.ValidateVolumes(spec.Volumes, fldPath.Child("volumes"))
```
We need to do something for validation here.  Can we easily use the swagger spec from the core API server?

pmorie commented on pkg/registry/settings/podpreset/storage.go:
```diff
@@ -0,0 +1,115 @@
+
+// NewSingular returns a new shell of a podpreset, according to the given namespace and
+// name
+func NewSingular(ns, name string) runtime.Object {
```
Do we need this if we don't have TPR support for this resource?

pmorie commented:
@droot 

What happens when I try to run this using TPR storage?  Does it fail?  Does it tell me that I need etcd for the settings API group?

droot commented:
@pmorie 
> What happens when I try to run this using TPR storage? Does it fail? Does it tell me that I need etcd for the settings API group?

Currently, the behavior is unknown. I have a TODO to handle this gracefully, will address it in this PR. Thanks for the comments.

jpeeler commented:
Closing as this is what https://github.com/kubernetes-incubator/service-catalog/pull/1694 was based off.
